fe/fe-core/src/main/java/com/starrocks/sql/analyzer/SelectAnalyzer.java

```diff
@@ -340,8 +341,13 @@ private void analyzeWhere(Expr whereClause, AnalyzeState analyzeState, Scope sco
         AnalyzerUtils.verifyNoWindowFunctions(predicate, "WHERE");
         AnalyzerUtils.verifyNoGroupingFunctions(predicate, "WHERE");
 
-        if (!predicate.getType().matchesType(Type.BOOLEAN) && !predicate.getType().matchesType(Type.NULL)) {
-            throw new SemanticException("WHERE clause must evaluate to a boolean: actual type %s", predicate.getType());
+
+        if (predicate.getType().isBoolean() || predicate.getType().isNull()) {
+            // do nothing
+        } else if (!session.getSessionVariable().isEnableStrictType() && Type.canCastTo(predicate.getType(), Type.BOOLEAN)) {
+            predicate = new CastExpr(Type.BOOLEAN, predicate);
+        } else {
+            throw new SemanticException("WHERE clause %s can not be converted to boolean type", predicate.toSql());
```

Comment:
add in ImplicitCastRule

Reply:
The reason for not modifying in the plan phase is that the analyze phase can more easily determine whether an expression is an expression in the where clause. For example, in `select 1 from tbl where 1`, the analyze phase can easily only change `where 1` to `where cast(1 as boolean)` but do noting to `select 1`